Add a PRAGMA to retrieve version

itsWill opened the pull request:
Fix: https://github.com/cwida/duckdb/issues/672

This adds a way for users to retrieve the version of duck db using SQL.
The possible methods which are all equivalent are:

`select * from  pragma_version();`

`PRAGMA version;`

These should return

library_version | source_id
DuckDB          | 76khli0

Theres another way to fetch the version using the inbuilt `.version`
function. This will go through the sqlite.c file and return the same
information, but it will also return the compiler version.

That is something that could also be added. Another thing we could add
to the output would be the git tag.

I'm proposing a minimal change set that we can iterate on.

Commit:
Add a PRAGMA to retrive version

Fix: https://github.com/cwida/duckdb/issues/672

This adds a way for users to retrieve the version of duck db using SQL.
The possible methods which are all equivalent are:

`select * from  pragma_version();`

`PRAGMA version;`

These should return

library_version | source_id
DuckDB          | 76khli0

Theres another way to fetch the version using the inbuilt `.version`
function. This will go through the sqlite.c file and return the same
information, but it will also return the compiler version.

That is something that could also be added. Another thing we could add
to the output would be the git tag.

I'm proposing a minimal change set that we can iterate on.

itsWill commented on src/function/table/CMakeLists.txt:
```diff
@@ -1,11 +1,13 @@
-                  copy_csv.cpp
-                  read_csv.cpp
-                  sqlite_functions.cpp)
+add_library_unity(
```
This change is the result of running `make format`. 

I'm ok with not keeping cosmetic changes in a PR, however the next person who runs `make format` on this file, won't also have to include this change if we keep it.

itsWill commented on src/function/table/sqlite/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
-                  pragma_collations.cpp
-                  pragma_table_info.cpp
-                  sqlite_master.cpp)
+add_definitions(-DDUCKDB_SOURCE_ID="\""${GIT_COMMIT_HASH}"\"")
```
I struggled a bit with this one, this is coming from the `tools/sql_api_wrapper` files. I tried to get a reference to the variable defined there, but my `CMake` skills are still poor and I couldn't, so I ended up copy pasting.

itsWill commented on src/function/table/sqlite/CMakeLists.txt:
```diff
@@ -1,8 +1,7 @@
-                  sqlite_master.cpp)
+add_definitions(-DDUCKDB_SOURCE_ID="\""${GIT_COMMIT_HASH}"\"")
+
+add_library_unity(duckdb_func_sqlite OBJECT pragma_collations.cpp
```
similar sentiment: https://github.com/cwida/duckdb/pull/808/files#r464251164

itsWill commented on src/function/table/sqlite/pragma_version.cpp:
```diff
@@ -0,0 +1,39 @@
+	}
+	output.SetCardinality(1);
+	output.SetValue(0, 0, "DuckDB");
+	output.SetValue(1, 0, DUCKDB_SOURCE_ID);
```
this is defined in the `CMakeLists.txt` file, howe it's also lives in the `sqlite_api_wrapper.cpp` file: https://github.com/cwida/duckdb/blob/7621c51e48bec80b61bede7f5fc5729ec4d09b76/tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp#L694-L699

Originally I had tried to `include <sqlite3.h>` and use the `sqlite3_libversion` and `sqlite3_source_id` functions. The linker couldn't find the `sqlite3_libversion` and `sqlite3_source_id` symbols, I tried to change various CMakeLists.txt to make it work, but I couldn't find quite the right incantation, so I ended up duplicating the functionality here.  Ideally we would be able to use those functions.

itsWill commented on test/sql/pragma/test_pragma_version.test:
```diff
@@ -0,0 +1,15 @@
+select * from pragma_version();
+
+query I
+select library_version from pragma_version();
```
It would be nice to test, the source_id column as well, but I think that would change on every commit making this test fail, a imperfect approximation could be to check that the source_id column is not null, or matches a regex of alphanumeric characters of a certain length.

Mytherin commented:
Thanks for the PR! I think it looks good, but indeed the duplication should be avoided. I would rather tackle this from the other end though: make the sqlite3_api_wrappers use the version found inside the DuckDB database, instead of the other way around.

How about this: we implement two static methods on the DuckDB class.

`static const char* DuckDB::LibraryVersion();`
`static const char* DuckDB::SourceID();`

The actual implementation of those functions should be in the file `src/function/table/sqlite/pragma_version.cpp`. Then the sqlite3_api_wrappers.cpp can call those methods in the `sqlite3_libversion` and `sqlite3_sourceid` methods. This is also useful because then other APIs (e.g. the Python or R API) can also easily get this version number.

For the define, it is important that the define only happens in a single CMake file that doesn't have many source files. This is because the define changes on every `git commit` and every `git pull`, which means any files that include this define need to be recompiled. We had this problem before where this define was global and the whole system would need to be recompiled on a `git commit` - we would like to avoid this. 

For that reason - I suggest we make a separate folder (`src/function/table/version`) that contains **only** the `pragma_version.cpp` and nothing else. Then only that function will need to be recompiled on a git commit hash change.

The test looks fine - perhaps an "IS NOT NULL" can be added, but otherwise it looks good.

Thanks again!

Commit:
Add LiobraryVersion and SourceID funcs to DuckDB

The implmenentation of these functions live in pragma_version.cpp the
sqlite3_api_wrapper file then calls out to these functions instead of
implementing them. This allows other tools like the rest/server.cpp
to also do the same.

Another advantage is that removing the GIT_HASH definition from various
CMake files into a single one in the table/sqlite/version folder means
less files have to be recompiled when the source changes on pulls or
commits.

itsWill commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -82,37 +79,37 @@ int sqlite3_open_v2(const char *filename, /* Database filename (UTF-8) */
-		*ppDb = nullptr;
-		if (zVfs) { /* unsupported so if set we complain */
-			return SQLITE_ERROR;
+		filename = NULL;
```
The changes in this function are al because of a `make format` run.

itsWill commented on tools/sqlite3_api_wrapper/sqlite3_api_wrapper.cpp:
```diff
@@ -692,10 +689,10 @@ void sqlite3_interrupt(sqlite3 *db) {
 
 const char *sqlite3_libversion(void) {
-	return LIB_VERSION;
+	return DuckDB::LibraryVersion();
```
These are non formatting changes reference the new functions.

itsWill commented:
@Mytherin great approach, I implemented it in: https://github.com/cwida/duckdb/pull/808/commits/e9bebefa126dcb2d7d126a6ff3fd13de774c7437 cleaned up some of the places that were referencing the `DUCKDB_SOURCE_ID` directly, to call this func.

 I've moved the `pragma_version.cpp` into the `table/version` folder as suggested, and consequently  I removed most of the `add_definition(--DDUCK_SOURCE_ID...)` calls, there's one that remains [here]( https://github.com/cwida/duckdb/blob/master/tools/shell/CMakeLists.txt#L3), I don't think it's used so I think we can remove it from there as well.

Mytherin commented:
Thanks for the changes! It looks a lot better now. 

> ...there's one that remains here, I don't think it's used so I think we can remove it from there as well.

Yes, please remove that one as well!

One last note: it seems that [the CI is failing](https://travis-ci.org/github/cwida/duckdb/jobs/714570375), it looks like a forgotten reference to `DUCKDB_SOURCE_ID`. Could you have a look?

After that is resolved it looks ready to merge from my side.

Commit:
add DUCKDB_SOURCE_ID definition to amalgamation.py

The DUCKDB_SOURCE_ID definition isn't defined from the CMakeLists like
when doing a `make debug` so I added a equivalent function to
amalgamation.py to define the DUCKDB_SOURCE_ID in the amalgamated file.

Commit:
remove -DDUCKDB_SOURCE_ID definition

Now that the pragma_version.cpp file implements the method
`DuckDB::SourceID()` we can use that instead of having to define the
variable in the CMakeLists.txt

In sqlite_api_wrapper.cc the method sqlite3_source_id is delegated to
the the `DuckDB::SourceID()` method, so we use that method in the
sqlite3.h header file to expose the source_id.

itsWill commented:
> Yes, please remove that one as well!

✅ 

> One last note: it seems that the CI is failing, it looks like a forgotten reference to DUCKDB_SOURCE_ID. Could you have a look?

You know what, it looks like it's happening because of the `amalgamation.py` script doesn't define the definitions in the `CMakeLists.txt` file, so I ended up adding a line there to define the `DUCK_DB_SOURCE` id var, in the script, in this [commit](https://github.com/cwida/duckdb/pull/808/commits/3b592c3e6f77f083eba086ea6ada919ee2696f18f)

Mytherin commented:
> You know what, it looks like it's happening because of the amalgamation.py script doesn't define the definitions in the CMakeLists.txt file, so I ended up adding a line there to define the DUCK_DB_SOURCE id var, in the script, in this commit

Thanks for looking into it! That's a great solution. I will wait until the tests pass and then merge it in; I think everything looks good now.
